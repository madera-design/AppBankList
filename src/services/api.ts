import axios from 'axios';

const API_URL = 'https://bptapidevmng.azure-api.net/catom/api/challenge/banks';

export const fetchBanks = async () => {
  try {
    const response = await axios.get(API_URL);
    return response.data;
  } catch (error) {
    console.error('Error fetching banks:', error);
    return [];
  }
};
