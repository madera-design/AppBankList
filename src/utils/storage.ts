import AsyncStorage from '@react-native-async-storage/async-storage';

export const saveBanks = async (banks: any) => {
  try {
    await AsyncStorage.setItem('banks', JSON.stringify(banks));
  } catch (error) {
    console.error('Error saving banks:', error);
  }
};

export const getSavedBanks = async () => {
  try {
    const data = await AsyncStorage.getItem('banks');
    return data ? JSON.parse(data) : null;
  } catch (error) {
    console.error('Error retrieving banks:', error);
    return null;
  }
};
