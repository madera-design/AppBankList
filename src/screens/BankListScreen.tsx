import React, { useEffect, useState } from 'react';
import { View, FlatList, ActivityIndicator, StyleSheet, Text } from 'react-native';
import { fetchBanks } from '../services/api';
import { saveBanks, getSavedBanks } from '../utils/storage';
import BankItem from '../components/BankItem';
import { Bank } from '../types/Bank';

const BankListScreen = () => {
    const [banks, setBanks] = useState<Bank[]>([]);
    const [loading, setLoading] = useState(true);
  
    useEffect(() => {
      const loadBanks = async () => {
        const cachedBanks = await getSavedBanks();
        if (cachedBanks) {
          setBanks(cachedBanks);
          setLoading(false);
        } else {
          const data: Bank[] = await fetchBanks();
          saveBanks(data);
          setBanks(data);
          setLoading(false);
        }
      };
  
      loadBanks();
    }, []);
  
    if (loading) {
      return (
        <View style={styles.loadingContainer}>
          <ActivityIndicator
            testID="loading-indicator"
            size="large" 
            color="#4A90E2" 
        />
        </View>
      );
    }
  
    return (
      <View style={styles.container}>
        <Text style={styles.header}>Lista de Bancos</Text>
        <FlatList
          data={banks}
          keyExtractor={(item) => item.bankName}
          renderItem={({ item }) => <BankItem bank={item} />}
        />
      </View>
    );
  };
  
  const styles = StyleSheet.create({
    container: {
      flex: 1,
      padding: 20,
      backgroundColor: '#f0f4f8',
    },
    loadingContainer: {
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: '#f0f4f8',
    },
    header: {
      fontSize: 24,
      fontWeight: 'bold',
      textAlign: 'center',
      marginBottom: 15,
      color: '#333',
    },
  });
  
  export default BankListScreen;
  
