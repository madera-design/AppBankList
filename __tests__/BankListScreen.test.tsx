import React from 'react';
import { render, waitFor } from '@testing-library/react-native';
import BankListScreen from '../src/screens/BankListScreen';
import { fetchBanks } from '../src/services/api';
import { saveBanks, getSavedBanks } from '../src/utils/storage';

// Mock de las funciones de la API y el almacenamiento
jest.mock('../src/services/api', () => ({
  fetchBanks: jest.fn(),
}));

jest.mock('../src/utils/storage', () => ({
  saveBanks: jest.fn(),
  getSavedBanks: jest.fn(),
}));

describe('BankListScreen Component', () => {
  const mockBanks = [
    {
      bankName: 'Banco 1',
      description: 'Descripci贸n 1',
      age: 5,
      url: 'https://example.com/logo1.png',
    },
    {
      bankName: 'Banco 2',
      description: 'Descripci贸n 2',
      age: 10,
      url: 'https://example.com/logo2.png',
    },
  ];

  it('shows loading indicator while fetching data', () => {
    const { getByTestId } = render(<BankListScreen />);
    expect(getByTestId('loading-indicator')).toBeTruthy();
  });

  it('renders list of banks after fetching data', async () => {
    // Simula que la API devuelve datos
    (fetchBanks as jest.Mock).mockResolvedValue(mockBanks);
    (getSavedBanks as jest.Mock).mockResolvedValue(null);

    const { getByText, queryByTestId } = render(<BankListScreen />);

    // Espera a que desaparezca el indicador de carga
    await waitFor(() => expect(queryByTestId('loading-indicator')).toBeNull());

    // Verifica que los bancos se rendericen
    expect(getByText('Banco 1')).toBeTruthy();
    expect(getByText('Descripci贸n 1')).toBeTruthy();
    expect(getByText('Banco 2')).toBeTruthy();
    expect(getByText('Descripci贸n 2')).toBeTruthy();
  });

  it('matches snapshot', async () => {
    (fetchBanks as jest.Mock).mockResolvedValue(mockBanks);
    (getSavedBanks as jest.Mock).mockResolvedValue(null);

    const { toJSON } = render(<BankListScreen />);
    await waitFor(() => expect(toJSON()).toMatchSnapshot());
  });
});