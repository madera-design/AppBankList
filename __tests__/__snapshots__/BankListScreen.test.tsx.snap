// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BankListScreen Component matches snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#f0f4f8",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    color="#4A90E2"
    size="large"
    testID="loading-indicator"
  />
</View>
`;
